doc/design/error_handling_requirements.md

```diff
@@ -0,0 +1,228 @@
+but do not necessarily terminate if it is not (depending on the `error_level`).
+If it is not true we raise an error of the required level.
+
+Hence
+`EXPECT(condition, error_code, FATAL, handler, optional_handler_args)`
+
+is equivalent to
+
+`ASSERT(condition, error_code, handler, optional_handler_args)`.
+
+If we do not need to call a handler and just provide some error message we can define
+
+`EXPECT(condition, error_code, error_level, error_message)`
+
+or even omit it entirely
+
+`EXPECT(condition, error_code, error_level)`
+
+We have the possibility to specify at compile time the minimum error level we intend to handle to e.g. exclude warnings.
+This can be done without causing overhead for the affected checks. Assuming there are the levels
+`WARNING`, `RECOVERABLE_ERROR`, `FATAL` we can suppress all checks for `WARNING` in a way that not even
+`condition` is evaluated in
+
+`EXPECT(condition, error_code, WARNING)`,
+
+i.e. achieve a zero-overhead abstraction. This is useful if `WARNING`s are indeed no problem that compromise the
+ability to continue but instead unusal system states we encounter.
+
+### Replacement of Expects and Ensures
+
+`ASSERT` should replace `Expects` and `Ensures`. In particular there is no use to distinguish between them unless it is
+possible to check the conditions **before** the function is actually invoked or immediately **after** it returns.
+It is not possible with the capabilities of the current C++ language to specify true preconditions and postconditions
+conveniently (e.g. without wrapping all functions manually or with some automation),
+hence both are basically custom asserts.
+
+This is consistent with the C++ error handling proposal <https://doc.bccnsoft.com/docs/cppreference2018/en/cpp/language/attributes/contract.html>
+which is encourages to use `assert` inside a function body.
+The proposed `ASSERT` is more general except in one regard: it cannot be disabled by design. Disabling it could easily
+be made possible, but this would lead to problems if thefailure conditions where ever encountered to be true
+since it would be impossible to safely proceed.
+
+## Error Propagation
+
+Idiomatic propagation of errors with `expected`, `optional` or return codes is not in scope here ad will be covered elsewhere.
+
+## Open Points
+
+1. Design considerations of functionality such as general handler code provided by the user
+1. Optional arguments if the proposed macro API path is chosen
+1. Replacement strategy and definition of the active error handler will be similar to the new logger (platforms specific)
+1. Prototype implementation and detailed design
+1. Error propataion with `expected` is another issue but does not concern the error handler invocation
+1. Replacement of noexcept to allow exceptions in client code (e.g. types, callbacks) is not in scope here
+1. Macro naming should be close to Google Test but ideally not the same to ease searches.
+   Currently it is just a rough proposal.
+
```

Comment:
At the end some food for thoughts.

I like the idea to call this `RAISE_something`. What do you think about aligning the error levels with the log levels?

This would mean that the macro would look like this `RAISE(FATAL, error code)`, `RAISE(ERROR, error code)` and `RAISE(WARNING, error code)`. Maybe with an `IOX_` prefix. With this it would be obvious which log level is used for the custom message. `ERROR` would be our current `SEVERE` and `WARNING` our current `MODERATE`.

Additionally there would be `RAISE_IF(condition, error level, error code)` and `ASSERT` would be syntactic sugar for this macro.